salt/states/x509.py

```diff
@@ -362,10 +366,112 @@ def csr_managed(name,
 
     return ret
 
+def _certificate_info_matches(cert_info, required_cert_info, check_serial=False):
+    '''
+    Return true if the provided certificate information matches the
+    required certificate information, i.e. it has the required common
+    name, subject alt name, organization, etc.
+
+    cert_info should be a dict as returned by x509.read_certificate.
+    required_cert_info should be a dict as returned by x509.create_certificate with testrun=True.
+    '''
+    # don't modify the incoming dicts
+    cert_info = copy.deepcopy(cert_info)
+    required_cert_info = copy.deepcopy(required_cert_info)
+
+    ignored_keys = [
+        'Not Before',
+        'Not After',
+        'MD5 Finger Print',
+        'SHA1 Finger Print',
+        'SHA-256 Finger Print',
+        # The integrity of the issuer is checked elsewhere
+        'Issuer Public Key'
+    ]
+    for key in ignored_keys:
+        cert_info.pop(key, None)
+        required_cert_info.pop(key, None)
+
+    if not check_serial:
+        cert_info.pop('Serial Number', None)
+        required_cert_info.pop('Serial Number', None)
+        try:
+            cert_info['X509v3 Extensions']['authorityKeyIdentifier'] = (
+                re.sub(r'serial:([0-9A-F]{2}:)*[0-9A-F]{2}', 'serial:--',
+                       cert_info['X509v3 Extensions']['authorityKeyIdentifier']))
+            required_cert_info['X509v3 Extensions']['authorityKeyIdentifier'] = (
+                re.sub(r'serial:([0-9A-F]{2}:)*[0-9A-F]{2}', 'serial:--',
+                       required_cert_info['X509v3 Extensions']['authorityKeyIdentifier']))
+        except KeyError:
+            pass
+
+    diff = []
+    for k, v in six.iteritems(required_cert_info):
+        try:
+            if v != cert_info[k]:
+                diff.append(k)
+        except KeyError:
+            diff.append(k)
+
+    return len(diff) == 0, diff
+
+
+def _certificate_days_remaining(cert_info):
+    '''
+    Get the days remaining on a certificate, defaulting to 0 if an error occurs.
+    '''
+    try:
+        expiry = cert_info['Not After']
+        return (
+            datetime.datetime.strptime(expiry, '%Y-%m-%d %H:%M:%S') -
+            datetime.datetime.now()).days
+    except KeyError:
+        return 0
+
+
+def _certificate_is_valid(name, days_remaining, append_certs, **cert_spec):
+    '''
+    Return True if the given certificate file exists, is a certificate, matches the given specification, and has the required days remaining.
+
+    If False, also provide a message explaining why.
+    '''
+    if not os.path.isfile(name):
+        return False, '{0} does not exist'.format(name), {}
+
+    try:
+        cert_info = __salt__['x509.read_certificate'](certificate=name)
+        required_cert_info = __salt__['x509.create_certificate'](testrun=True, **cert_spec)
```

Comment:
It looks like this still creates new certificates before it knows whether they're needed (and then creates another one later the new one is needed)?
We had discussed this before in #52167.

Reply:
I'm sorry @OrangeDog, I've misunderstood #52167 - I thought it was to do with the certificate comparisons (e.g. new cert created even though the requested `days_remaining` is `14` and the existing cert is `21`), not the `copypath` on the CA server filling up with certs.

I'm not going to attempt to fix #52167 here to keep the scope of this PR reasonable. Sorry! 

`x509.create_certificate` with `testrun=True` is designed to create a cert in memory that we can compare against here in the state. We should be able to call it with `testrun=True` as much as we need without filling up `copypath`. It's a bug in that execution module, which I'd like to fix in another PR if this one ever gets merged. I've updated the description of this PR to reflect this.